File: composeApp/src/androidMain/kotlin/com/example/showdown26/ui/components/PokemonAvatar.kt
package com.example.showdown26.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.example.showdown26.R

@Composable
fun PokemonAvatar(
    imageUrl: String?,
    contentDescription: String?,
    circleSize: Dp,
    spriteSize: Dp,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier.size(spriteSize),
        contentAlignment = Alignment.Center
    ) {
        PokeballCircle(modifier = Modifier.size(circleSize))
        PreviewAsyncImage(
            url = imageUrl,
            previewDrawable = R.drawable.preview_pokemon,
            contentDescription = contentDescription,
            modifier = Modifier.size(spriteSize)
        )
    }
}

@Composable
fun FillPokemonAvatar(
    imageUrl: String?,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    circleFraction: Float = 0.7f
) {
    Box(
        modifier = modifier,
        contentAlignment = Alignment.Center
    ) {
        PokeballCircle(modifier = Modifier.fillMaxSize(circleFraction))
        PreviewAsyncImage(
            url = imageUrl,
            previewDrawable = R.drawable.preview_pokemon,
            contentDescription = contentDescription,
            modifier = Modifier.fillMaxSize()
        )
    }
}

@Composable
private fun PokeballCircle(modifier: Modifier = Modifier) {
    val themeColor = MaterialTheme.colorScheme.primary

    Canvas(modifier = modifier) {
        val strokeWidth = size.minDimension * 0.04f
        val inset = strokeWidth / 2
        val drawSize = Size(size.width - strokeWidth, size.height - strokeWidth)
        val center = Offset(size.width / 2, size.height / 2)
        val radius = drawSize.minDimension / 2
        val topLeft = Offset(inset, inset)

        // Top half (theme color)
        drawArc(
            color = themeColor,
            startAngle = 180f,
            sweepAngle = 180f,
            useCenter = true,
            topLeft = topLeft,
            size = drawSize
        )

        // Bottom half (white)
        drawArc(
            color = Color.White,
            startAngle = 0f,
            sweepAngle = 180f,
            useCenter = true,
            topLeft = topLeft,
            size = drawSize
        )

        // Horizontal line
        drawLine(
            color = Color.Black,
            start = Offset(inset, center.y),
            end = Offset(size.width - inset, center.y),
            strokeWidth = strokeWidth
        )

        // Circle outline
        drawCircle(
            color = Color.Black,
            radius = radius,
            center = center,
            style = Stroke(width = strokeWidth)
        )

        // Center button
        val buttonRadius = radius * 0.24f
        drawCircle(color = Color.White, radius = buttonRadius, center = center)
        drawCircle(
            color = Color.Black,
            radius = buttonRadius,
            center = center,
            style = Stroke(width = strokeWidth)
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun PokemonAvatarPreview() {
    MaterialTheme {
        PokemonAvatar(
            imageUrl = null,
            contentDescription = "Dragonite",
            circleSize = 100.dp,
            spriteSize = 144.dp
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun FillPokemonAvatarPreview() {
    MaterialTheme {
        FillPokemonAvatar(
            imageUrl = null,
            contentDescription = "Dragonite",
            modifier = Modifier.size(100.dp)
        )
    }
}

File: iosApp/iosApp/PokemonAvatar.swift
import SwiftUI

let isPreview = ProcessInfo.processInfo.environment["XCODE_RUNNING_FOR_PREVIEWS"] == "1"

// MARK: - Silhouette Modifier

extension View {
    /// When `useSilhouettes` is enabled, renders the view as a solid black shape.
    @ViewBuilder
    func ifSilhouette() -> some View {
        if useSilhouettes {
            self
                .foregroundStyle(.black)
                .colorMultiply(.black)
        } else {
            self
        }
    }
}

// MARK: - Theme Color Environment Key

private struct ThemeColorKey: EnvironmentKey {
    static let defaultValue: Color = Color(red: 0.863, green: 0.184, blue: 0.208) // Red
}

extension EnvironmentValues {
    var themeColor: Color {
        get { self[ThemeColorKey.self] }
        set { self[ThemeColorKey.self] = newValue }
    }
}

// MARK: - Pokeball Circle

struct PokeballCircle: View {
    @Environment(\.themeColor) var themeColor

    var body: some View {
        Canvas { context, size in
            let strokeWidth = min(size.width, size.height) * 0.04
            let inset = strokeWidth / 2
            let drawRect = CGRect(x: inset, y: inset, width: size.width - strokeWidth, height: size.height - strokeWidth)
            let center = CGPoint(x: size.width / 2, y: size.height / 2)
            let radius = min(drawRect.width, drawRect.height) / 2

            // Top half (theme color)
            var topPath = Path()
            topPath.addArc(center: center, radius: radius, startAngle: .degrees(180), endAngle: .degrees(0), clockwise: false)
            topPath.closeSubpath()
            context.fill(topPath, with: .color(themeColor))

            // Bottom half (white)
            var bottomPath = Path()
            bottomPath.addArc(center: center, radius: radius, startAngle: .degrees(0), endAngle: .degrees(180), clockwise: false)
            bottomPath.closeSubpath()
            context.fill(bottomPath, with: .color(.white))

            // Horizontal line
            var linePath = Path()
            linePath.move(to: CGPoint(x: center.x - radius, y: center.y))
            linePath.addLine(to: CGPoint(x: center.x + radius, y: center.y))
            context.stroke(linePath, with: .color(.black), lineWidth: strokeWidth)

            // Circle outline
            context.stroke(Path(ellipseIn: drawRect), with: .color(.black), lineWidth: strokeWidth)

            // Center button
            let buttonRadius = radius * 0.24
            let buttonRect = CGRect(x: center.x - buttonRadius, y: center.y - buttonRadius, width: buttonRadius * 2, height: buttonRadius * 2)
            context.fill(Path(ellipseIn: buttonRect), with: .color(.white))
            context.stroke(Path(ellipseIn: buttonRect), with: .color(.black), lineWidth: strokeWidth)
        }
    }
}

// MARK: - Pokemon Avatar

struct PokemonAvatar: View {
    let imageUrl: String?
    let circleSize: CGFloat
    let spriteSize: CGFloat

    var body: some View {
        ZStack {
            if useSilhouettes {
                Circle()
                    .fill(Color(.systemGray5))
                    .frame(width: circleSize, height: circleSize)
            } else {
                PokeballCircle()
                    .frame(width: circleSize, height: circleSize)
            }
            PreviewAsyncImage(url: imageUrl, previewAsset: "PreviewPokemon")
                .frame(width: spriteSize, height: spriteSize)
                .ifSilhouette()
        }
        .frame(width: spriteSize, height: spriteSize)
    }
}

struct FillPokemonAvatar: View {
    let imageUrl: String?
    var circleFraction: CGFloat = 0.7

    var body: some View {
        GeometryReader { geometry in
            let size = geometry.size.width
            let circleSize = size * circleFraction

            ZStack {
                if useSilhouettes {
                    Circle()
                        .fill(Color(.systemGray5))
                        .frame(width: circleSize, height: circleSize)
                } else {
                    PokeballCircle()
                        .frame(width: circleSize, height: circleSize)
                }
                PreviewAsyncImage(url: imageUrl, previewAsset: "PreviewPokemon")
                    .frame(width: size, height: size)
                    .ifSilhouette()
            }
            .frame(width: size, height: size)
        }
        .aspectRatio(1, contentMode: .fit)
    }
}

#Preview("PokemonAvatar") {
    PokemonAvatar(imageUrl: nil, circleSize: 100, spriteSize: 144)
}

#Preview("FillPokemonAvatar") {
    FillPokemonAvatar(imageUrl: nil)
        .frame(width: 100, height: 100)
}
